From: Gael Varoquaux <gael.varoquaux@normalesup.org>
Subject: BUG: fix change of behavior in last commit

--- a/sklearn/metrics/metrics.py
+++ b/sklearn/metrics/metrics.py
@@ -524,9 +524,19 @@ def roc_curve(y_true, y_score, pos_label
 
     if not (n_pos is np.nan or n_neg is np.nan):
         # add (0,0) and (1, 1)
-        fpr = np.r_[0., fpr, 1.]
-        tpr = np.r_[0., tpr, 1.]
-        thresholds = np.r_[thresholds[0] + 1, thresholds, thresholds[-1] - 1]
+        if not (fpr[0] == 0 and fpr[-1] == 1):
+            fpr = np.r_[0., fpr, 1.]
+            tpr = np.r_[0., tpr, 1.]
+            thresholds = np.r_[thresholds[0] + 1, thresholds,
+                               thresholds[-1] - 1]
+        elif not fpr[0] == 0:
+            fpr = np.r_[0., fpr]
+            tpr = np.r_[0., tpr]
+            thresholds = np.r_[thresholds[0] + 1, thresholds]
+        elif not fpr[-1] == 1:
+            fpr = np.r_[fpr, 1.]
+            tpr = np.r_[tpr, 1.]
+            thresholds = np.r_[thresholds, thresholds[-1] - 1]
     elif fpr.shape[0] == 2:
         # trivial decisions, add (0,0)
         fpr = np.array([0.0, fpr[0], fpr[1]])
